### PR TITLE
Logical Restructure of Alerting plugin servers

### PR DESCRIPTION
**Pull request by alexandreLamarre**
Part of https://github.com/rancher/opni/issues/1153

Moving alerting server implementations into discrete structs

### CODE REVIEW

File: pkg/util/errors.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+)
+
+// IErrGroup
+// Best effort to run all tasks, then combines all errors
+type IErrGroup struct {
```
Comment (kralicky) [inline pkg/util/errors.go:12]: Suggest renaming, this doesn't match go naming conventions.

File: pkg/util/init.go [hunk 2]
```diff
@@ -58,3 +58,10 @@ func (i *Initializer) checkInitCond() {
 		i.initCond = sync.NewCond(&sync.Mutex{})
 	})
 }
+
+type InitializerF interface {
```
Comment (kralicky) [inline pkg/util/init.go:62]: This should probably stay within your plugin

**Review by kralicky — CHANGES_REQUESTED** [review]

**Review by dbason — APPROVED** [review]

**Review by kralicky — CHANGES_REQUESTED** [review]

File: plugins/alerting/pkg/alerting/alarms/v1/status.go [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+			State:  alertingv1.AlertConditionState_Invalidated,
+			Reason: "cluster does not have metrics capabilities installed",
+		}
+	}
+	if !cond.GetLastUpdated().AsTime().Before(time.Now().Add(-(60 * time.Second))) {
+		return &alertingv1.AlertStatusResponse{
+			State:  alertingv1.AlertConditionState_Pending,
+			Reason: "alarm metric dependencies are updating",
+		}
+	}
```
[the remark this answers is not in the document]
Comment (alexandreLamarre) [reply plugins/alerting/pkg/alerting/alarms/v1/status.go:84]: rules take an evaluation period to run

File: pkg/util/errors.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+	i.WaitGroup.Wait()
+}
+
+func (i *MultiErrGroup) AddError(err error) {
+	i.errMu.Lock()
+	defer i.errMu.Unlock()
+	i.errs = append(i.errs, err)
+}
+
+func (i *MultiErrGroup) Error() error {
```
[the remark this answers is not in the document]
Comment (alexandreLamarre) [reply pkg/util/errors.go:36]: I think `Combine` is better

File: plugins/alerting/pkg/alerting/alarms/v1/status.go [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+		State: alertingv1.AlertConditionState_Ok,
+	}
+}
+
+func (a *AlarmServerComponent) loadStatusInfo(ctx context.Context) (*statusInfo, error) {
+	// lg := p.Logger.With("request", "loadStatusInfo")
+
+	ctxca, ca := context.WithTimeout(ctx, 10*time.Second)
+	defer ca()
+	eg, _ := errgroup.WithContext(ctxca)
```
[the remark this answers is not in the document]
Comment (alexandreLamarre) [reply plugins/alerting/pkg/alerting/alarms/v1/status.go:117]: no

File: plugins/alerting/pkg/alerting/alarms/v1/status.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexandreLamarre) [reply plugins/alerting/pkg/alerting/alarms/v1/status.go:83]: it is technically configurable, but here its used as a heuristic

[the remark this answers is not in the document]
Comment (alexandreLamarre) [reply plugins/alerting/pkg/alerting/alarms/v1/status.go:83]: two things have to happen consecutively for the rule state to be accurate from cortex:

- a completed poll interval for persisted rule configs that checks if there are persisted configs that need to be loaded as prometheus rules. this is controlled via an independent poll config duration.
- a poll interval for loaded rules in memory that evaluates them w.r. to discovered metrics, one full evaluation is required to bring the rule from an unknown state to ok, firing, error, etc...

this is an opinionated heuristic where we omit handling a varierty of weird edge-case statuses in the setup phase, while also not passing on details of the setup implementation to the end user

**Review by kralicky — APPROVED** [review]

**Pull request merged**